### ISSUE TITLE
Inner build does not correctly require Drill

### ISSUE THREAD

**Issue by xsebek** [swarm-game/swarm#394]
From #373 by @byorgey:

> So I ran into one additional issue while testing out some things. In creative mode I tried something like this:
> ```
> def forever = \c. c ; forever c end
> def unblock = try {drill forward} {} end
> def push = unblock; move end
> 
> build {forever (push; build {turn right; forever push})}
> ```
> ...
> However, it doesn't work. The initial robot works fine, drilling along and building another robot at each step. However, all those secondary robots don't end up with a drill device and crash with a fatal error when they try to execute the drill command.

The problem was present before, but not visible until the robots stopped without a drill.

In this new iron age, it is immediately a fatal bug once the robot attempts to drill.

## Screenshots

A bronze age screenshot from the main branch:

![Screenshot from 2022-06-14 17-01-12](https://user-images.githubusercontent.com/44544735/173629200-67c3adf3-3af3-4f2b-9f41-3940b1d4a359.png)
<sup>It's almost the same now too, so I won't include another screenshot here.</sup>

Log of the robot missing a drill:

![Screenshot from 2022-06-14 17-19-14](https://user-images.githubusercontent.com/44544735/173629709-3bdb3bce-fb4d-420f-aca5-a912561010b9.png)

## Additional context

I traced the issue to the `requiredCaps` function which did not require `CDrill`.

**Comment by byorgey** [comment]
Oh, I like your use of `log` and `robotNamed` to be able to find and view the log for the inner robot. =D

**Comment by xsebek** [comment]
Surprisingly this program in Classic mode does not run into a Fatal error:
```
def forever = \c. c ; forever c end;
def unblock = try {drill forward} {} end;
def push = unblock; move end;
log "Hi, I am base";
r <- build {
  wait 2;
  log "Hi, I am builder";
  forever (build {log "Hi, I am pusher"; turn forward; forever push}; log "- robot built")
};
give r "logger";
give r "drill";
```
These are all the robots logs:
```
Hi, I am base
Hi, I am builder
- robot built
build: You do not have required devices, please obtain: 
  - logger
Hi, I am pusher
You do not have the devices required for:
  'drill'
  please install:
   - metal drill or drill
```

**Comment by xsebek** [comment]
Once #400 is merged, you can test this with an included scenario:
```bash
$ cabal run test:swarm-integration -- -p '/build with drill (#394)/'
Up to date
Tests
  Test scenario solutions
    Regression tests
      build with drill (#394): FAIL (expected: Awaiting fix (#394)) (0.04s)
        test/integration/Main.hs:155:
        Fatal error: Drill is required but not installed?!
        Please report this as a bug at
        <https://github.com/swarm-game/swarm/issues/new>.
         (expected failure)

All 1 tests passed (0.04s)
```

**Comment by xsebek** [comment]
@byorgey is this #540?

**Comment by byorgey** [comment]
No, it's not #540.  That one happens when you have a `def` and a `build` in the same expression, separated by a semicolon.  You can tell that's not the problem since the `push` command *does* correctly lead to a `drill` being required for the outer robot, just not for the inner robots.  However, it could be that somehow the correct requirements context with `push` in it is not being passed along to the `requireCaps` function for the inner `build`.  I am still not sure.

**Comment by byorgey** [comment]
OK, so the issue seems to be that newly built robots were not inheriting the `robotContext` of their parent, so the inner robot did not know what the requirements were for `push`.  Copying the `robotContext` over is easy enough, but first I want to understand why a lot of other things weren't already broken by this.  Like, why are built robots still able to use definitions from the parent robot, if they have an empty context?

**Comment by byorgey** [comment]
Hmm, this actually gives me an idea how to fix #540...
